newton.go

```diff
@@ -0,0 +1,203 @@
+// direction. ModifiedNewton implements a Hessian modification strategy that
+// tries to add successively larger multiples of identity to H_k until it
+// becomes positive definite. Note that the repeated trial factorization of the
+// modified Hessian involved in this process can be computationally expensive.
+// For points in the vicinity of a minimizer ModifiedNewton reduces to a pure
+// Newton's method with its quadratic convergence rate.
+//
+// If the Hessian matrix cannot be formed explicitly or if the computational
+// cost of its factorization is prohibitive, BFGS or L-BFGS quasi-Newton method
+// can be used instead.
+type ModifiedNewton struct {
+	// LinesearchMethod is a method used for selecting suitable steps along the
+	// descent direction d. Steps should satisfy the Wolfe, Goldstein or Armijo
+	// conditions. If LinesearchMethod == nil, an appropriate default is
+	// chosen.
+	LinesearchMethod LinesearchMethod
+	// Increase is a factor by which a scalar tau is successively increased so
+	// that (H + tau*I) is positive definite. Larger values reduce the number
+	// of trial Hessian factorizations, but also reduce the second-order
+	// information in H.
+	// Increase must be greater than 1. If Increase is 0, it is defaulted to 5.
+	Increase float64
+
+	linesearch *Linesearch
+
+	hess *mat64.SymDense
+	chol *mat64.TriDense
+	tau  float64
+}
+
+func (n *ModifiedNewton) Init(loc *Location, f *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if n.Increase == 0 {
+		n.Increase = 5
+	}
+	if n.Increase <= 1 {
+		panic("optimize: ModifiedNewton.Increase must be greater than 1")
+	}
+	if n.LinesearchMethod == nil {
+		n.LinesearchMethod = &Bisection{}
+	}
+	if n.linesearch == nil {
+		n.linesearch = &Linesearch{}
+	}
+	n.linesearch.Method = n.LinesearchMethod
+	n.linesearch.NextDirectioner = n
+
+	return n.linesearch.Init(loc, f, xNext)
+}
+
+func (n *ModifiedNewton) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return n.linesearch.Iterate(loc, xNext)
+}
+
+func (n *ModifiedNewton) InitDirection(loc *Location, dir []float64) (stepSize float64) {
+	dim := len(loc.X)
+	n.chol = resizeTriDense(n.chol, dim)
+	n.hess = resizeSymDense(n.hess, dim)
+	n.tau = 0
+	n.computeNextDir(loc, dir)
+	return 1
+}
+
+func (n *ModifiedNewton) NextDirection(loc *Location, dir []float64) (stepSize float64) {
+	n.computeNextDir(loc, dir)
+	return 1
+}
+
+func (n *ModifiedNewton) computeNextDir(loc *Location, dir []float64) {
+	dim := len(loc.X)
+	n.hess.CopySym(loc.Hessian)
+
+	// Find the smallest diagonal element of H.
+	minA := n.hess.At(0, 0)
+	for i := 1; i < dim; i++ {
+		a := n.hess.At(i, i)
+		if a < minA {
+			minA = a
+		}
+	}
+	if minA > 0 {
+		// All diagonal elements are positive, so try to factorize non-modified
+		// Hessian first.
+		n.tau = 0
+	} else if n.tau == 0 {
+		// Use the last tau that gave PD modified Hessian.
```

Comment:
Oh! The comment is misleading. The comment in the block applies to the missing "else", not to the code right there. I think it's better to have a comment above the block. Something like
````
// If the smallest diagonal entry is positive, the Hessian may be positive definite, and so first attempt an un-modified cholesky. If the smallest entry is negative, use the final tau from the last iteration if regularization was needed, otherwise guess an appropriate value for tau
````